### PR TITLE
feat(sdk): Rewind

### LINKED ISSUE #1395: Overwriting old history
Is it possible to overwrite old history with wandb. I have use-case, when my experiment diverged and I lowered learning rate and used older checkpoint to restart experiment, but wandb doesn’t overwrite history and shows me this warning:
```WARNING Adding to old History rows isn't currently supported.```

I was wondering maybe there is flag to overwrite old history (didn't find in documentation)? If there is no such flag, is it possible to add this as feature request?
Thank you in advance!

### PR DESCRIPTION
**Pull request by dannygoldstein**
Description
-----------

- Fixes WB-5045
- Fixes #1395  
- Addresses many other issues

This PR implements the SDK changes needed for "run rewinding," a long requested feature that allows users to truncate the history logged to a run at an intermediate step and resume logging from that step. It uses run forking under the hood. 

The API is similar to forking but it uses a different init() kwarg,

```python
resumed_run = wandb.init(resume_from='[old_run_id]?_step=[desired_step]')
```

This does some metadata shifting to get the behavior requested in https://github.com/wandb/wandb/issues/1395. 
- [x] I updated CHANGELOG.md, or it's not applicable

Testing
-------
How was this PR tested?

Tested manually as this is gated behind the v2 runs store at the moment so cannot be effectively or meaningfully unit tested, same as in #7078.

### CODE REVIEW

File: wandb/sdk/wandb_run.py
Comment (kptkin) [inline wandb/sdk/wandb_run.py:709]: maybe make the setting oneof? so we guranteed we don't have both fork_from and resume_from?

Comment (kptkin) [reply wandb/sdk/wandb_run.py:709]: or i guess now looking at the type, maybe just a filed that says what type of xx_from?

File: wandb/sdk/wandb_init.py
Comment (kptkin) [inline wandb/sdk/wandb_init.py:1161]: i guess you could replace it with a set and check the length the same way

File: wandb/sdk/wandb_settings.py
Comment (kptkin) [inline wandb/sdk/wandb_settings.py:1876]: should it be in it's own function? as these not really the same right?

File: wandb/sdk/wandb_settings.py
Comment (kptkin) [inline wandb/sdk/wandb_settings.py:1883]: is it ok that the run_id could be overwritten later on in the function?

File: wandb/sdk/internal/internal_api.py
Comment (kptkin) [inline wandb/sdk/internal/internal_api.py:2230]: could you provide a better explantation for this function?

File: wandb/sdk/internal/internal_api.py
Comment (kptkin) [inline wandb/sdk/internal/internal_api.py:2286]: what happens for servers that don't support this feature?

File: wandb/sdk/internal/sender.py
Comment (kptkin) [inline wandb/sdk/internal/sender.py:916]: what if the step is 0 and we didn't log anything does it mean we will have an offset by 1?

File: wandb/sdk/internal/sender.py
Comment (kptkin) [inline wandb/sdk/internal/sender.py:918]: could you leave a comment here, since it is under resume but you mark `forked` as `True`?

File: wandb/sdk/internal/sender.py
Comment (kptkin) [inline wandb/sdk/internal/sender.py:897]: why do you need to keep this one as part of the sender class?

File: wandb/sdk/wandb_run.py
Comment (dannygoldstein) [reply wandb/sdk/wandb_run.py:709]: see convo here - https://weightsandbiases.slack.com/archives/C0108SLEP6K/p1717518143092499?thread_ts=1717466352.551879&cid=C0108SLEP6K

i implemented some logic in the message deserializer that keeps the message structure the same (for backcompat purposes and separation of concerns) and ensures that we never instantiate a settings object in the backend that has both of these specified

File: wandb/sdk/wandb_settings.py
Comment (dannygoldstein) [reply wandb/sdk/wandb_settings.py:1883]: i dont think it could right? because on the only other subsequent `self.update` line that could overwrite, the argument is `self.run_id or generate_id()`, but since we set `self.run_id` on this line, that subsequent line will always be a no op

File: wandb/sdk/wandb_settings.py
Comment (dannygoldstein) [reply wandb/sdk/wandb_settings.py:1876]: factored it out

File: wandb/sdk/internal/internal_api.py
Comment (dannygoldstein) [reply wandb/sdk/internal/internal_api.py:2230]: Oof, sorry. This was AI generated ... fixed.

File: wandb/sdk/internal/internal_api.py
Comment (dannygoldstein) [reply wandb/sdk/internal/internal_api.py:2286]: we will just error here. is there a better practice that you recommend?

File: wandb/sdk/internal/sender.py
Comment (dannygoldstein) [reply wandb/sdk/internal/sender.py:916]: yeah this is a good question. i guess in the current formulation the step is exclusive. we could make it inclusive - would that be more intuitive?

Comment (dannygoldstein) [reply wandb/sdk/internal/sender.py:916]: i think it makes sense to be exclusive for fork, but maybe it should be inclusive for rewind

File: wandb/sdk/internal/sender.py
Comment (dannygoldstein) [reply wandb/sdk/internal/sender.py:918]: done

File: wandb/sdk/internal/sender.py
Comment (dannygoldstein) [reply wandb/sdk/internal/sender.py:897]: because it needs to be accessed by `_init_run()` and I didn't want to change the function signature of that method.

File: wandb/sdk/internal/sender.py
Comment (dannygoldstein) [reply wandb/sdk/internal/sender.py]: made this inclusive (i.e., the situation you describe above will no longer happen)

File: wandb/sdk/wandb_init.py
Comment (timoffex) [inline wandb/sdk/wandb_init.py:1160]: `kwargs = dict(locals())` is an anti-pattern that we should be avoiding and are trying to remove.

```suggestion
    num_resume_options_set = (
        (fork_from is not None)  # wrap
        + (resume is not None)
        + (resume_from is not None)
    )
```

File: wandb/sdk/wandb_settings.py
Comment (timoffex) [inline wandb/sdk/wandb_settings.py:1863]: this comment doesn't seem to serve a purpose; it's just the function name?

File: wandb/sdk/wandb_settings.py
Comment (timoffex) [inline wandb/sdk/wandb_settings.py:1864]: should invert `if` statement to reduce nesting

```python
if self.resume_from is None:
    return

...
```

File: wandb/sdk/internal/sender.py
Comment (timoffex) [inline wandb/sdk/internal/sender.py:1088]: it's not clear from the code why `_setup_rewind()` is not always called when `is_rewinding` is true

File: wandb/sdk/internal/internal_api.py
Comment (timoffex) [inline wandb/sdk/internal/internal_api.py:2239]: super nit: some of these don't end with a period, whereas others do. They should all consistently end with a period.

I'm also not sure whether argument types are necessary in this docstring, since they're already present in the function signature. This is an older style that existed before type annotations I believe. @kptkin do you know if this affects the generated docs?

File: wandb/sdk/internal/internal_api.py
Comment (timoffex) [inline wandb/sdk/internal/internal_api.py:2240]: The docstring should document its return value as well. It should either enumerate what one expects to be in the dictionary or write down how one can find out (by referencing some other documentation, maybe?)

**Review by timoffex — COMMENTED** [review]
Just a few style comments, but otherwise LGTM. I might not be aware of some SDK nuances, so I'll let @kptkin approve when it's all cleared up.

File: wandb/sdk/internal/internal_api.py
Comment (kptkin) [reply wandb/sdk/internal/internal_api.py:2239]: good question. not 100% but i think we need that with this format. cc: @ngrayluna to confirm

File: wandb/sdk/wandb_settings.py
Comment (dannygoldstein) [reply wandb/sdk/wandb_settings.py:1863]: fixed

File: wandb/sdk/wandb_settings.py
Comment (dannygoldstein) [reply wandb/sdk/wandb_settings.py:1864]: fixed

File: wandb/sdk/internal/sender.py
Comment (dannygoldstein) [reply wandb/sdk/internal/sender.py]: added comment

File: wandb/sdk/internal/internal_api.py
Comment (dannygoldstein) [reply wandb/sdk/internal/internal_api.py:2239]: This is the same format as for `upsert_run`

File: wandb/sdk/internal/internal_api.py
Comment (dannygoldstein) [reply wandb/sdk/internal/internal_api.py:2261]: done - but this is not done in upsert_run currently

File: wandb/sdk/internal/sender.py
Comment (dannygoldstein) [reply wandb/sdk/internal/sender.py]: on second thought I think this needs to match the behavior in fork, because the keysInfo and history recomputation on the backend assume exclusivity here. But I think the example provided is a bit contrived. There is not much use rewinding to zero, that is effectively the same as starting a new run. So there is a clear workaround in that case - just call `wandb.init()` normally

File: wandb/sdk/internal/internal_api.py
Comment (kptkin) [reply wandb/sdk/internal/internal_api.py:2307]: did you try to test it? will it crush the entire script or hang?

File: wandb/sdk/internal/internal_api.py
Comment (kptkin) [inline wandb/sdk/internal/internal_api.py:2314]: do you need to use get here? in case the response is empty?

File: wandb/sdk/internal/sender.py
Comment (timoffex) [inline wandb/sdk/internal/sender.py:1037]: we have the lint for a reason; either the original code should be refactored or the additions should be written such that we don't need to disable the "function is too complex" check

File: wandb/sdk/internal/internal_api.py
Comment (timoffex) [reply wandb/sdk/internal/internal_api.py:2261]: thank you!

> this is not done in upsert_run currently

it's not a status quo that we should strive to keep :)

File: wandb/sdk/internal/internal_api.py
Comment (dannygoldstein) [reply wandb/sdk/internal/internal_api.py:2314]: nice catch. done

File: wandb/sdk/internal/sender.py
Comment (dannygoldstein) [reply wandb/sdk/internal/sender.py]: i refactored it

**Review by kptkin — APPROVED** [review]

**Comment by JosselinSomervilleRoberts** [comment]
For anyone having issues with this feature but still needing it, here is my hacky solution:
* Start your runs with `wandb.init(name=UUID, resume=None, ...)`. Do not use `id`
* This will create a new run every time you relaunch/your job gets preempted which is not what we want BUT you can get a satisfying result if:
    - every time you log using `step=`. At this point you should get something like this:
image
    - This is not what we want but at least we have both runs with the actual iterations done by each launch. Now you can simply go to "Runs", Select "Group" and choose "Name".  
image
Here I also went to edit the plot and under grouping chose "Range=Samples". It is very dim but in the overlap section you see in dotted lines the values for both runs. The actual line will show the average, min or max depending on what you choose. Sadly there is no way to show the latest but that's already decent for my use case.

Hopefully this will help some of you!

**Pull request merged**
